### PR TITLE
Simplify policy code

### PR DESCRIPTION
**Pull request by MindFlavor**
This PR tries to make the policy code easier to follow by:

1. Limit the number of retry policies to 1 for pipeline. As seen in the C# SDK ([https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#azure-storage](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#azure-storage)) the retry policy is set at client level, and it is limited to 1 per pipeline. Detangling the retry policy from the pipeline also allows to optionally specify a custom retry policy per request (as it's possible in the C# sdk).
2. Making the pipeline idempotent. This is critical since in case of errors the pipeline must restart with the same request.
3. Fixing the number of TransportPolicies to 1 and forcing it to be the last policy. 

This PR also implements (taken from [https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-mechanism-10](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-mechanism-10)):

1. Linear backoff retry policy.
2. Exponential backoff retry policy.
3. No retry policy.

### CODE REVIEW

File: sdk/core/src/policy.rs [hunk 1]
```diff
@@ -118,30 +68,68 @@ impl Context {
...
+    async fn send(&self, ctx: Context, request: Request) -> PolicyResult<R>;
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Pipeline {
-    policies: Vec<Arc<dyn Policy>>,
+    policies: Vec<Arc<dyn Policy<Request>>>,
+    retry: Arc<dyn RetryPolicy>,
+    transport: Arc<dyn Policy<Response>>,
```
Comment (rylev) [inline sdk/core/src/policy.rs:82]: I believe we also want to break the policies into pre-retry and post-retry policies. Can we do that now?

File: sdk/core/src/policy.rs [hunk 2]
```diff
@@ -1,58 +1,15 @@
-use crate::{Request, Response};
-
-use async_trait::async_trait;
-
+use crate::retry_policy::RetryPolicy;
+use crate::Request;
+use crate::Response;
```
Comment (rylev) [inline sdk/core/src/policy.rs:3]: Nit: why did we break this into two lines?

File: sdk/core/src/policy.rs [hunk 3]
```diff
@@ -118,30 +68,68 @@ impl Context {
...
-        next: &[Arc<dyn Policy>],
-    ) -> PolicyResult<Response>;
+pub trait Policy<R>: Send + Sync + std::fmt::Debug
+where
+    R: Send + Sync,
+{
+    async fn send(&self, ctx: Context, request: Request) -> PolicyResult<R>;
 }
 
-#[derive(Debug, Clone)]
```
Comment (rylev) [inline sdk/core/src/policy.rs:130]: If we can't derive `Debug` we should have a custom impl for it.

File: sdk/core/src/request.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+use http::{HeaderMap, Method};
+use std::fmt::Debug;
+use url::Url;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Request {
+    pub url: Url,
+    pub method: Method,
+    pub headers: HeaderMap,
+    pub payload: bytes::Bytes,
```
Comment (rylev) [inline sdk/core/src/request.rs:10]: I would rather request be completely opaque. This helps with future compatibility. Can we just make these getters and setters instead?

File: sdk/core/src/request.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+impl Request {
+    #[allow(dead_code)]
```
Comment (rylev) [inline sdk/core/src/request.rs:14]: Why?

File: sdk/core/src/request.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
+use http::{HeaderMap, Method};
```
Comment (rylev) [inline sdk/core/src/request.rs:1]: I'd like to eventually get rid of any visible dependency on `http`. It would be nice to move in that direction.

File: sdk/core/src/retry_policy.rs [hunk 7]
```diff
@@ -0,0 +1,123 @@
+use tokio::time::{sleep, Duration};
+
+#[async_trait::async_trait]
+pub trait RetryPolicy {
```
Comment (rylev) [inline sdk/core/src/retry_policy.rs:4]: nit: I think we should get in the habit of documenting things so we don't make the lack of documentation worse than it is currently.

File: sdk/core/src/retry_policy.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    async fn create_instance(&self) -> Box<dyn RetryPolicyInstance>;
```
Comment (rylev) [inline sdk/core/src/retry_policy.rs:5]: Is there a reason `create_instance` is async?

File: sdk/core/src/retry_policy.rs [hunk 9]
```diff
@@ -0,0 +1,123 @@
+use tokio::time::{sleep, Duration};
```
Comment (rylev) [inline sdk/core/src/retry_policy.rs:1]: It's a real shame that we'll end up relying on tokio, but I guess that's the way it will have to be...

File: sdk/core/src/retry_policy.rs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+#[async_trait::async_trait]
+pub trait RetryPolicy {
+    async fn create_instance(&self) -> Box<dyn RetryPolicyInstance>;
+}
+
+#[async_trait::async_trait]
+pub trait RetryPolicyInstance {
+    async fn retry(&mut self) -> bool;
```
Comment (rylev) [inline sdk/core/src/retry_policy.rs:10]: nit: the name `retry` makes it sound like this is in charge of retrying. Perhaps `should_retry` would be clearer?

File: sdk/core/src/request.rs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    pub(crate) fn into_parts(self) -> (Parts, bytes::Bytes) {
+        let parts = Parts {
+            url: self.url,
+            method: self.method,
+            headers: self.headers,
+        };
+        (parts, self.payload)
+    }
+
+    pub fn set_payload(&mut self, bytes: bytes::Bytes) {
```
Comment (rylev) [inline sdk/core/src/request.rs:24]: Is the plan to eventually allow for non-bytes related payloads(e.g., streams)?

File: sdk/core/src/request.rs (hunk 4 above)
Comment (rylev) [inline sdk/core/src/request.rs:10]: I'm not sure I like the name `payload` better than `body`.

File: sdk/core/src/policy.rs [hunk 12]
```diff
@@ -118,30 +68,68 @@ impl Context {
...
 
 #[async_trait::async_trait]
-pub trait Policy: Send + Sync + std::fmt::Debug {
-    async fn send(
-        &self,
-        ctx: Context,
-        request: &mut Request,
-        next: &[Arc<dyn Policy>],
-    ) -> PolicyResult<Response>;
+pub trait Policy<R>: Send + Sync + std::fmt::Debug
```
Comment (rylev) [inline sdk/core/src/policy.rs:71]: I'm confused why `Policy` is generic...

**Review by rylev — CHANGES_REQUESTED** [review]
Good start! Not sure I agree that this is "simplifying" the policy code, but it does add functionality.

File: sdk/core/src/policy.rs (hunk 1 above)
Comment (MindFlavor) [reply sdk/core/src/policy.rs:82]: Sorry I missed that. This means that we should have infallible policies that are executed only after a retry? I am not clear on this point.

File: sdk/core/src/policy.rs (hunk 2 above)
Comment (MindFlavor) [reply sdk/core/src/policy.rs:3]: No reason! Will fix!

File: sdk/core/src/request.rs (hunk 4 above)
Comment (MindFlavor) [reply sdk/core/src/request.rs:10]: Yes I agree. What do you think about the `getset` crate?

File: sdk/core/src/request.rs (hunk 5 above)
Comment (MindFlavor) [reply sdk/core/src/request.rs:14]: Since we don't have - yet - any other custom policy mutability is not needed. I've left `mut` there to signal that the vector should be `mut` but it's not (temporarily). I can remove it!

File: sdk/core/src/retry_policy.rs (hunk 7 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:4]: Yes, at least I should document the logic behind the traits because it's expected for clients to implement their own retry policies (at least C# SDK allows it).

File: sdk/core/src/retry_policy.rs (hunk 8 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:5]: In our case no but since clients could implement `RetryPolicy` I think it's a good idea to allow them to use async calls. For example you can have a `RetryPolicy` that "resets" a connection (for example to a db) and it could definitely require an async call.

File: sdk/core/src/request.rs (hunk 11 above)
Comment (MindFlavor) [reply sdk/core/src/request.rs:24]: Yes, it's something I was working on! Also I would like to update the response to better expose the return stream if need be.

File: sdk/core/src/policy.rs (hunk 12 above)
Comment (MindFlavor) [reply sdk/core/src/policy.rs:71]: The idea is that a Policy can return either an updated Request (for example the CredentialsPolicy will add the authorization headers) or a Response. 

I was wondering if returning **always** a stream from TransportPolicy can be a good solution. We could then add an additional policy called `CollectStreamPolicy` that transforms the stream into a `bytes::Bytes`. Only functions that need to return the collected stream will add this additional policy. For example, for create database, the pipeline could be:

1. Prepare headers + method + body (specific of create database).
2. `CredentialsPolicy` adds auth headers (or bearer token).
3. `TelemetryPolicy` will add telemetry headers.
3. `TransportPolicy` *spends* the request and gives back a stream.
4. `CollectStreamPolicy` converts the stream into `bytes::Bytes`.
5. A custom `CreateDatabasePolicy` parses the stream and generates a `CreateDatabaseResponse`.

This way the entire create database code will be expressed in policies (most of which are reused across the crates). 

Does it make sense?

File: sdk/core/src/policy.rs (hunk 1 above)
Comment (rylev) [reply sdk/core/src/policy.rs:82]: I believe there are certain policies which should *not* be redone *before* the retry policy and the retry policy only retries policies that come after it in the pipeline.

File: sdk/core/src/request.rs (hunk 4 above)
Comment (rylev) [reply sdk/core/src/request.rs:10]: I'd rather avoid a dependency if we can. Let's write it out and see how hard it is to maintain. It would be trivial to write a macro_rules macro for this case, if we really need it.

File: sdk/core/src/retry_policy.rs (hunk 8 above)
Comment (rylev) [reply sdk/core/src/retry_policy.rs:5]: But wouldn't that resetting of the connection happen in the `retry` method of `RetryPolicyInstance`?

File: sdk/core/src/policy.rs (hunk 12 above)
Comment (rylev) [reply sdk/core/src/policy.rs:71]: Hmm.... The idea was that policies return the response. If the policy is receiving an owned `Request` why would it need to return the `Request`? This seems to complicate things. Perhaps you could show an example where this allows for flexibility where the non-generic trait could not handle.

File: sdk/core/Cargo.toml [hunk 13]
```diff
@@ -35,9 +35,9 @@ oauth2 = "=4.0.0-alpha.3"
 reqwest = "0.11"
 paste = "1.0"
 md5 = "0.7"
+tokio = "1.0"
```
Comment (ctaggart) [inline sdk/core/Cargo.toml:38]: Tokio does not support WebAssembly. I'll probably look at seeing what it takes to get core and the services to target wasm soon.

File: sdk/core/src/policy.rs [hunk 14]
```diff
@@ -1,58 +1,15 @@
-use crate::{Request, Response};
-
-use async_trait::async_trait;
-
+use crate::retry_policy::RetryPolicy;
```
Comment (heaths) [inline sdk/core/src/policy.rs:1]: While you're at it - or I can later - could you take a look at my open draft PR to refactor specific policies into another mod? There will be more, and this is similar to how we organize them all in our other SDKs to avoid monolithic files.

File: sdk/core/src/policy.rs (hunk 1 above)
Comment (heaths) [reply sdk/core/src/policy.rs:82]: That's correct. All our SDKs are fairly consistent on this, though JS's recent changes are a little more robust with scheduling. But to keep thing simple now, see an example at https://github.com/Azure/azure-sdk-for-net/blob/7e3cf643977591e9041f4c628fd4d28237398e0b/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L35.

File: sdk/core/src/policy.rs [hunk 15]
```diff
@@ -118,30 +68,68 @@ impl Context {
...
 
-    pub async fn send(&self, ctx: Context, mut request: Request) -> PolicyResult<Response> {
-        self.policies[0]
-            .send(ctx, &mut request, &self.policies[1..])
-            .await
+    pub async fn send(&self, ctx: Context, request: Request) -> PolicyResult<Response> {
+        // with create instance we make sure every pipeline execution has a new retry policy
+        // instance (with for example the counters reset to zero) based off the template policy
+        // specified in the pipeline.
+        let mut retry_policy = self.retry.create_instance().await;
```
Comment (heaths) [inline sdk/core/src/policy.rs:102]: I think we should keep this as a RetryPolicy. Policies are supposed to be fully customizable even to the point of removing or rescheduling where this retry is performed. Making everything a policy right up to and including the TransportPolicy has been working well for all the other SDKs.

File: sdk/core/src/retry_policy.rs [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+            .await;
+            true
+        } else {
+            false
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ExponentialBackoffRetryPolicy {
```
Comment (heaths) [inline sdk/core/src/retry_policy.rs:61]: I don't see why this is a separate policy. We have a single RetryPolicy across other SDKs that uses service-provided back-offs if supported, or just uses linear. Making individual clients decide or even customers just shifts the problem/responsibility down to developers when we can handle it upstream in core.

**Review by heaths — CHANGES_REQUESTED** [review]

File: sdk/core/src/policy.rs (hunk 14 above)
Comment (MindFlavor) [reply sdk/core/src/policy.rs:1]: Good idea! 👍

File: sdk/core/Cargo.toml (hunk 13 above)
Comment (MindFlavor) [reply sdk/core/Cargo.toml:38]: [http://www.nooooooooooooooo.com/](http://www.nooooooooooooooo.com/)

I'll sketch a thread sleep and remove tokio.

File: sdk/core/src/retry_policy.rs (hunk 16 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:61]: I see. Good point. 
I'll remove the ExponentialBackoff and just leave linear and no retry policy.

Comment (heaths) [reply sdk/core/src/retry_policy.rs:61]: > and no retry policy

Not sure I follow. We should have retries as a policy as mentioned above. It makes customizing the policies in the pipeline straight forward with a consistent debugging pattern.

File: sdk/core/src/retry_policy.rs (hunk 8 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:5]: Yes it could happen there. I would have preferred to separate the initialization phase of the policy instance from its execution (like a constructor), but I suppose I would have to make it fallible as well.

Let's make it simple and keep it infallible and sync! :+1:

File: sdk/core/src/retry_policy.rs (hunk 16 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:61]: mmm now I'm confused. In the [docs](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#retry-mechanism-10) we say:

> The built-in classes provide support for __linear__ (constant delay) and __exponential__ with randomization retry intervals. There is also a __no retry__ policy for use when another process is handling retries at a higher level. However, you can implement your own retry classes if you have specific requirements not provided by the built-in classes.

So I thought to code the very same policies here. 

Do you mean that we should have a single, customizable policy instead of 3 different policies?

File: sdk/core/src/retry_policy.rs (hunk 9 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:1]: I've removed the dependency because I've implemented a `sleep` future by hand (it just spawns a thread that wakes the reactor after sleeping). For the time being I think it is good enough.

File: sdk/core/src/request.rs (hunk 4 above)
Comment (MindFlavor) [reply sdk/core/src/request.rs:10]: renamed!

File: sdk/core/Cargo.toml (hunk 13 above)
Comment (MindFlavor) [reply sdk/core/Cargo.toml:38]: Done!

File: sdk/core/src/retry_policy.rs (hunk 16 above)
Comment (MindFlavor) [reply sdk/core/src/retry_policy.rs:61]: > Making individual clients decide or even customers just shifts the problem/responsibility down to developers when we can handle it upstream in core.

At least for C#, this is definitely the case: [https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.tablerequestoptions.retrypolicy?view=azure-dotnet#Microsoft_Azure_Cosmos_Table_TableRequestOptions_RetryPolicy](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.tablerequestoptions.retrypolicy?view=azure-dotnet#Microsoft_Azure_Cosmos_Table_TableRequestOptions_RetryPolicy). 
Developers can choose which retry policy they want (they are provided a default but that's it).

Also, looking at the implementations, they have the aforementioned retry policies: [https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.iretrypolicy?view=azure-dotnet](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.table.iretrypolicy?view=azure-dotnet):

* Microsoft.Azure.Cosmos.Table.ExponentialRetry
* Microsoft.Azure.Cosmos.Table.IExtendedRetryPolicy
* Microsoft.Azure.Cosmos.Table.LinearRetry
* Microsoft.Azure.Cosmos.Table.NoRetry

Arren't we supposed to copy this behaviour?

**Comment by MindFlavor** [comment]
In the last commit I've split the policies in perCall and in perRetry as done in the C# pipeline crate (see [https://github.com/Azure/azure-sdk-for-net/blob/7e3cf643977591e9041f4c628fd4d28237398e0b/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L35](https://github.com/Azure/azure-sdk-for-net/blob/7e3cf643977591e9041f4c628fd4d28237398e0b/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L35)).

The pipeline logic is now documented in the code since I'm not sure it is the right one I'll paste here to facilitate discussion:

The `Pipeline::send` function performs these steps:
1. For each `per_call_policy` execute it and pass the result to the following policy. In case of errors just return it (no retry).
2. Instantiate a retry policy instance with retry count to zero using the `retry_policy` as template.
3. For each `per_retry_policy` execute it and pass the result to the following policy. In case of errors, ask the retry policy if the pipeline should retry the policies. If yes, reset the request at the end of `per_call_policies` pipeline and execute the  `per_retry_policies` again.
4. Send the request to transport. In case of errors, behave like `per_retry_policy` invoking the `retry_policy` first then optionally retry.

The `context` can be mutated by each policy and is **not** reset between retries. It can be used to store information with the lifecycle of the whole pipeline execution.

File: sdk/core/src/context.rs [hunk 17]
```diff
@@ -0,0 +1,12 @@
+#[derive(Clone)]
+pub struct Context {
```
Comment (rylev) [inline sdk/core/src/context.rs:2]: Can we add documentation to the struct and the `new` function?

File: sdk/core/src/lib.rs [hunk 18]
```diff
@@ -30,16 +34,19 @@ use oauth2::AccessToken;
...
 pub use http_client::{to_json, HttpClient};
 pub use models::*;
-pub use policy::*;
+pub use request::*;
 
 pub type RequestId = Uuid;
 pub type SessionToken = String;
 pub const EMPTY_BODY: &[u8] = &[];
 
+type BoxedFuture<T> = Box<dyn std::future::Future<Output = policies::PolicyResult<T>> + Send>;
```
Comment (rylev) [inline sdk/core/src/lib.rs:48]: If we're hardcoding `PolicyResult` maybe this should be inside the `policy` module

File: sdk/core/src/pipeline.rs [hunk 19]
```diff
@@ -0,0 +1,96 @@
+use crate::policies::{Policy, PolicyResult, RetryPolicy};
+use crate::{Context, Request, Response};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct Pipeline {
```
Comment (rylev) [inline sdk/core/src/pipeline.rs:6]: Documentation

File: sdk/core/src/pipeline.rs [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+            f,
+            "Pipeline{{per_call_policies.len() == {}, per_retry_policies.len() == {}}}",
+            self.per_call_policies.len(),
+            self.per_retry_policies.len(),
+        )
+    }
+}
+
+impl Pipeline {
+    pub fn new(
```
Comment (rylev) [inline sdk/core/src/pipeline.rs:25]: Documentation

File: sdk/core/src/pipeline.rs [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    retry: Arc<dyn RetryPolicy>,
+    per_retry_policies: Vec<Arc<dyn Policy<Request>>>,
+    transport: Arc<dyn Policy<Response>>,
+}
+
+impl std::fmt::Debug for Pipeline {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Pipeline{{per_call_policies.len() == {}, per_retry_policies.len() == {}}}",
```
Comment (rylev) [inline sdk/core/src/pipeline.rs:17]: Not sure how helpful this debug info is. We can leave it for now though.

File: sdk/core/src/pipeline.rs [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+                        continue;
+                    } else {
+                        return Err(request_result.err().unwrap());
+                    }
+                } else {
+                    request = request_result.unwrap();
+                }
+            }
+
+            let response = self.transport.send(ctx, request.clone()).await;
```
Comment (rylev) [inline sdk/core/src/pipeline.rs:84]: Since we want to treat the transport policy as if it were a `per_retry_policy` we can just do `for policy in self.per_retry_policies.iter().chain(std::iter::once(&self.transport))`

File: sdk/core/src/policy.rs (hunk 12 above)
Comment (rylev) [reply sdk/core/src/policy.rs]: @MindFlavor I'm still not convinced that this architecture is the right way to go - having `Policy` be generic makes things more complicated than they need be.

File: sdk/core/src/policies/retry_policies/no_retry.rs [hunk 23]
```diff
@@ -0,0 +1,17 @@
+use crate::policies::{RetryPolicy, RetryPolicyInstance};
+
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct NoRetryPolicy {}
```
Comment (rylev) [inline sdk/core/src/policies/retry_policies/no_retry.rs:4]: nit: we might want to give this a private field so that users can't instantiate it directly. This would allow us to add fields in the future without breaking users. This might not be necessary, but better than being forced to not ever be able to add a field.

File: sdk/core/src/policies/transport.rs [hunk 24]
```diff
@@ -0,0 +1,49 @@
+use crate::policies::{Policy, PolicyResult};
+use crate::{BoxedFuture, Context, Request, Response};
+use std::pin::Pin;
+use std::sync::Mutex;
+
+type Transport = dyn Fn(&mut Context, Request) -> Pin<BoxedFuture<Response>> + Send;
```
Comment (rylev) [inline sdk/core/src/policies/transport.rs:6]: I do wonder if modeling this as a closure is the right thing to do. Another possibility is to have a trait like `Transport` which would have one method `send` on it. This might make implementation easier than a closure. Not sure...

File: sdk/core/src/request.rs [hunk 25]
```diff
@@ -0,0 +1,66 @@
+use http::{HeaderMap, Method};
+use std::fmt::Debug;
+use url::Url;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Request {
```
Comment (rylev) [inline sdk/core/src/request.rs:6]: Why was this moved to a separate module? If we don't want `Request` together with response than we should move `Response` to its own module. I'd prefer there to be an `http` module that `Request` and `Response` were both a part of.

File: sdk/core/src/request.rs [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+    pub url: Url,
+    pub method: Method,
+    pub headers: HeaderMap,
+    pub body: bytes::Bytes,
```
Comment (rylev) [inline sdk/core/src/request.rs:10]: Can you remove the `pub` on all these fields?

File: sdk/core/src/sleep.rs [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+        duration,
+    }
+}
+
+pub(crate) struct Sleep {
+    thread: Option<thread::JoinHandle<()>>,
+    duration: Duration,
+}
+
+impl Future for Sleep {
```
Comment (rylev) [inline sdk/core/src/sleep.rs:19]: Is implementing our own Sleep future the right way to go? This is fairly expensive (one thread per sleep). But perhaps this is the best way to go about this....

File: sdk/core/src/lib.rs (hunk 18 above)
Comment (MindFlavor) [reply sdk/core/src/lib.rs:48]: Defintely! 👍

File: sdk/cosmos/src/clients/cosmos_client.rs [hunk 28]
```diff
@@ -22,11 +24,18 @@ const TIME_FORMAT: &str = "%a, %d %h %Y %T GMT";
...
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 enum TransportStack {
     Client(Arc<Box<dyn HttpClient>>),
     Pipeline(Pipeline),
 }
+
+impl std::fmt::Debug for TransportStack {
+    fn fmt(&self, _: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        todo!()
```
Comment (rylev) [inline sdk/cosmos/src/clients/cosmos_client.rs:35]: Don't forget this!

File: sdk/cosmos/src/clients/cosmos_client.rs [hunk 29]
```diff
@@ -36,18 +45,18 @@ pub struct CosmosClient {
 }
 /// TODO
```
Comment (rylev) [inline sdk/cosmos/src/clients/cosmos_client.rs:46]: I know this was already here, but can you update this?

File: sdk/core/src/pipeline.rs (hunk 21 above)
Comment (MindFlavor) [reply sdk/core/src/pipeline.rs:17]: It's really not that useful. But I do not see how to make it more informative short of constraining `Policy` to be `Debug` as well 🤷‍♂️

**Review by rylev — CHANGES_REQUESTED** [review]
Looking better, but I'm still not sure about some things.

File: sdk/core/src/policies/retry_policies/exponential_retry.rs [hunk 30]
```diff
@@ -0,0 +1,78 @@
+use crate::policies::{RetryPolicy, RetryPolicyInstance};
+use crate::sleep::sleep;
+use chrono::{DateTime, Local};
+use std::time::Duration;
+
+/// Built in retry policies that support linear and exponential backoff (with an added random delay
```
[the remark this answers is not in the document]
Comment (MindFlavor) [reply sdk/core/src/policies/retry_policies/exponential_retry.rs:6]: Will do! 👍.

File: sdk/core/src/policies/transport.rs (hunk 24 above)
Comment (MindFlavor) [reply sdk/core/src/policies/transport.rs:6]: I am mulling over it as well. 
Closures are very succinct but can quickly become complicated to reason with. We'll probably have to reiterate more on this.

File: sdk/core/src/request.rs (hunk 25 above)
Comment (MindFlavor) [reply sdk/core/src/request.rs:6]: In the end request will be much more complicated because it will need to support AsyncRead + AsyncSeek streams as well as Bytes.  That's why I've elected to give it a separated module (but we should also do it for response of course 🤦‍♂️ ).

File: sdk/core/src/pipeline.rs (hunk 22 above)
Comment (MindFlavor) [reply sdk/core/src/pipeline.rs:84]: They are different types, per_retry_policies are `Request` -> `Request`, transport is `Request` -> `Response`  because `Policy` is basically `Request` -> `T` (or in Rust terms `Policy<Request>`, `Policy<Response>`). This brings us back to the generic in `Policy` 🤯 ...

BTW in C# the transport is not a policy at all (see [https://github.com/Azure/azure-sdk-for-net/blob/7e3cf643977591e9041f4c628fd4d28237398e0b/sdk/core/Azure.Core/api/Azure.Core.net5.0.cs#L228](https://github.com/Azure/azure-sdk-for-net/blob/7e3cf643977591e9041f4c628fd4d28237398e0b/sdk/core/Azure.Core/api/Azure.Core.net5.0.cs#L228)). It's just an abstract base class. 

I do not see why we cannot follow the C# sdk: if we make the transport a different, specific type we could get rid of the generic in `Policy` (and the signature would be `Request` -> `Request`). 

What do you think? I will try to add a sketch in the next PR so we can see if it works.

Comment (rylev) [reply sdk/core/src/pipeline.rs:84]: In the current architecture (what's in the master branch), policies are bit more powerful. Because they are responsible for calling into the policies collection, they can choose whether future policies get invoked or not. This is not possible with your changes. We need to check whether that sort of flexibility is needed.

File: sdk/core/src/request.rs (hunk 25 above)
Comment (rylev) [reply sdk/core/src/request.rs:6]: I fine with `Request` being in its own module, but I think then we should have an `http` module which has two submodules: `request` and `response`.

File: sdk/core/src/retry_policy.rs (hunk 16 above)
Comment (heaths) [reply sdk/core/src/retry_policy.rs]: Those are track 1 SDKs. For the most part, they are deprecated. Anything Microsoft.Azure.* is not worth looking at for guidance. Azure.* is. Our track 2 `RetryPolicy` is here: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Pipeline/Internal/RetryPolicy.cs

You'll also notice where *it* does the retry loop, not the pipeline. This would be consistent with all our languages, where request goes through the policy chain and responses come out, but the pipeline doesn't interact with it directly. You could, for example, even take retries out without having to have a `NoRetryPolicy`.

**Comment by MindFlavor** [comment]
Closing since the state machine approach proposed here does not conform to the other SDKs. 
I will open another PR keeping the recursive, mutating logic.
